### PR TITLE
Revise Broadcast reference implementation

### PR DESCRIPTION
**Pull request by mbencer**
This PR depends on #2641 (tile reference implementation) which should be merged first.

- improved Broadcast ref implementation
- reduced binary size by not using templates
- added single layer tests
- fixed handling boolean type for CPU plugin

### CODE REVIEW

File: ngraph/core/reference/src/runtime/reference/tile.cpp [hunk 1]
```diff
@@ -14,34 +14,103 @@
...
+    bool run = true;
+    std::vector<int64_t> indices(in_shape_expanded.size() - 1, 0);
+    size_t axis(indices.size());
+    const int input_rank = in_shape_expanded.size();
+    const int64_t last_dim = in_shape_expanded[input_rank - 1];
+    const std::vector<int64_t> pitches = create_pitches(out_shape);
+    const char* copy = nullptr;
 
-    for (const Coordinate& output_coord : output_transform)
+    while (run)
```
Comment (ilyachur) [inline ngraph/core/reference/src/runtime/reference/tile.cpp:88]: I really dislike such infinite cycles.
Looks like we can change the conditional (we change run only if `axis-- == 0`.

Comment (mbencer) [reply ngraph/core/reference/src/runtime/reference/tile.cpp:88]: @ilyachur agree, but I think you should add this comment to https://github.com/openvinotoolkit/openvino/pull/2641. From broadcast perspective I call only `reference::tile` (this PR depends on #2641)

File: inference-engine/tests/ngraph_functions/src/broadcast.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+//
+
+#include <vector>
+#include <memory>
+
+#include "ngraph_functions/builders.hpp"
+
+namespace ngraph {
+namespace builder {
+std::shared_ptr<ngraph::Node> makeBroadcast(const ngraph::Output<Node> &in,
```
Comment (ilyachur) [inline inference-engine/tests/ngraph_functions/src/broadcast.cpp:12]: If I remember right we already have make_broadcast function in the ngraph::builder static library. Why didn't you use this function?

**Review by tomdol — APPROVED** [review]

**Review by postrational — APPROVED** [review]

**Review by iefode — APPROVED** [review]
Test part LGTM

File: inference-engine/tests/ngraph_functions/src/broadcast.cpp (hunk 2 above)
Comment (mbencer) [reply inference-engine/tests/ngraph_functions/src/broadcast.cpp:12]: @ilyachur that's right, but `builder::opset1::make_broadcast` produces only `Broadcast::v1` in `EXPLICIT` mode.

Comment (ilyachur) [reply inference-engine/tests/ngraph_functions/src/broadcast.cpp:12]: Yeah, but why we cannot extend `make_broadcast` function? 
Also opset1 and opset5 are the same for Broadcast layer. We can bump the opset version in the broadcast builder.

**Review by ilyachur — APPROVED** [review]
LGTM

File: ngraph/core/reference/src/runtime/reference/tile.cpp [hunk 3]
```diff
@@ -52,53 +52,72 @@ void runtime::reference::tile(const char* arg,
...
-    const int64_t last_dim = in_shape_expanded[input_rank - 1];
-    const std::vector<int64_t> pitches = create_pitches(out_shape);
-    const char* copy = nullptr;
-
-    std::vector<int64_t> indices(in_shape_expanded.size() - 1, 0);
-    size_t axis = indices.size();
-
-    // Copy and repeat data for innermost axis as many times as described in the repeats parameter
-    while (axis <= indices.size())
+    try
```
Comment (ilyachur) [inline ngraph/core/reference/src/runtime/reference/tile.cpp:55]: Could you explain why you added this try/catch and changed the Tile code?

Should we add more tests for Tile op?

Comment (mbencer) [reply ngraph/core/reference/src/runtime/reference/tile.cpp:55]: @ilyachur I've observed exception in e2e after merge https://github.com/openvinotoolkit/openvino/pull/2715/commits/0c5e17d42ea196cf4a71f10aa44a8bccac1c1a8f with very limited error message. It is hard to reproduce it locally. This try/catch is only to collect info about error scenario and I'll revert it.
To the number of tests, probably right, we should add more

File: inference-engine/tests/ngraph_functions/src/broadcast.cpp (hunk 2 above)
Comment (mbencer) [reply inference-engine/tests/ngraph_functions/src/broadcast.cpp:12]: @ilyachur Broadcast:v3 was introduced in opset3 (with new `bidirectional` mode).
What's more `builder::opset1::make_broadcast` uses inverted logic of axes for explicit mode - take axes broadcasted in output (Broadcast:v0 logic), where Broadcast:v1,v2 take axes from input which will be broadcasted.
But you are right it should be refactored - I think that after removing legacy opset0

**Review by GlebKazantaev — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

**Comment by mbencer** [comment]
The PR is waiting for resolving problem with int8 transformation

**Pull request merged**